### ISSUE TITLE
DEP_FOO_KEY-like system that can work without "links"

### ISSUE THREAD

**Issue by dwrensha** [rust-lang/cargo#3544]
Would it make sense to allow usage of [`DEP_<name>_<key`](
http://doc.crates.io/environment-variables.html#environment-variables-cargo-sets-for-build-scripts) environment variables in a `build.rs` for dependencies that do *not* link to native libraries?

Context: I would like to allow crates like [this one](https://crates.io/crates/sandstorm) to expose Cap'n Proto schema definitions, and in particular to allow Cap'n Proto's [import mechanism](https://capnproto.org/language.html#imports) to work across crate boundaries. (See https://github.com/dwrensha/capnpc-rust/issues/30.)

For example, say I write a `my_app` crate that depends on the `sandstorm` crate, and I want `my_app` to have some capnp schemas that import from the `sandstorm` schemas. For this to work, the `my_app` `build.rs` needs to know where to find the source code of the `sandstorm` schemas. One way to accomplish that would be for the `sandstorm` crate's `build.rs` to write its current working directory to a `DEP_SANDSTORM_IMPORT` environment variable, through the mechanism described [here](http://doc.crates.io/build-script.html#the-links-manifest-key). However, that only works if the `sandstorm` crate has a "links" field in its `Cargo.toml`. Since the `sandstorm` crate does not actually need to link to a native library, adding such a field feels superfluous and possibly prone to mysterious bugs.

**Comment by alexcrichton** [comment]
Yeah I've often desired such a feature as well. I've had difficulty rationalizing in the past what `<name>` would be, however, to disambiguate from other crates. For example if two versions of the same library are linked into a crate how would those keys be passed to the build script? Cargo doesn't support this feature today, but it may wish to one day.

Other than that though it seems like a reasonable feature to me!

**Comment by sfackler** [comment]
You can't depend directly on two versions of the same crate though, right? Are `DEP_*` variables from transitive dependencies exposed to build scripts?

**Comment by alexcrichton** [comment]
@sfackler not currently, no, but I could imagine adding that as a feature to Cargo at some point. I think we'd only want to expose immediate dependencies to build scripts, not transitive, as that'd make the problem much easier (and more practical I think)

**Comment by alexcrichton** [comment]
Well so we may be able to actually just go ahead and do this.

So Cargo will pass a number of `--extern` flags with some name. That is *the* namespace for incoming crates. If Cargo ever supports multiple versions it'll have to support renaming crates anyway (e.g. renamed when passing to `--extern`. In that sense I think it's safe to say there's a namespace here anyway and it will fit naturally into the env vars.

We'd probably want to spec that the name in the env var though is the name of the library, not the name of the package. That differs in a few rare cases but would future-proof us a bit.

**Comment by retep998** [comment]
You currently can depend on two crates that have the same *library* name, yet different package names.

**Comment by SimonSapin** [comment]
I fully expected `<name>` to be the crate/package’s name, was confused for quite some time as to why by dependen’t build script did not get any `DEP_*` environment variable, and only finally figured it out by reading Cargo’s own test suite.

I’ve ended up adding to a crate a `links` key with a dummy value.

**Comment by retep998** [comment]
These `DEP_*` keys are the *only* way we have as far as I can tell to provide information to downstream consumers. The fact that `links` has to be specified to be able to use this feature is a real hassle, especially for crates that don't want cargo to limit them to only a single version (otherwise [this](https://github.com/retep998/winapi-rs/issues/238) happens).

**Comment by alexcrichton** [comment]
I'll again refer to my previous comments:

* [Not originally done](https://github.com/rust-lang/cargo/issues/3544#issuecomment-273243725) due to suspected name collisions.
* [Original thoughts seem incorrect](https://github.com/rust-lang/cargo/issues/3544#issuecomment-273865931), should be possible to implement!

@SimonSapin @retep998 if you'd like to send a PR for this it'd definitely be most welcome!

**Comment by retep998** [comment]
If we do add `DEP_*` keys following the *crate* name, what do we do when there are collisions with the existing `DEP_*` keys using the `links` value? Would we effectively declare the old way deprecated and only provide `DEP_*` keys using `links` when they don't conflict with a crate name?

**Comment by alexcrichton** [comment]
We could just use a different namespace, such as `CARGO_DEP_FOO_KEY=value`

**Comment by sfackler** [comment]
I think it'd be a good idea in general to start migrating to a `CARGO_` prefix on environment variables.

**Comment by indygreg** [comment]
I have a pretty similar case as the initial reporter for indygreg/PyOxidizer. I have a crate with a build script that needs to generate a library containing a special build of Python. The build process also generates some files which are consumed by callers into that library from a different crate. I'd like for the crate doing all the heavy lifting to expose the paths to these build artifacts to dependent crates so they can pick them up and use them.

I can kinda coerce the existing `DEP_*` variables to work with `links`, but it feels brittle since I'm not actually declaring a library dependency in the crate setting that key. That feels like the kind of thing that could break at any time since behavior sounds undefined.

**Comment by cr1901** [comment]
I probably need this functionality for a crate that doesn't depend on any native library. So I'm likely to go ahead and set `links` to `name-of-crate-$COOKIE`, where `$COOKIE` was set with `mcookie`.

This is [almost surely](https://twitter.com/incomputable/status/1623725557396525058) not going to conflict.

It's not great, but I'll put a comment back to this issue explaining the rationale.

**Comment by gzz2000** [comment]
I need this, too. I think it can be helpful to just kind of expose `CARGO_DEP_MANIFEST_DIR_{dep crate name}` and `CARGO_DEP_OUR_DIR_{dep crate name}` envs to dependents.

**Comment by abrown7100** [comment]
I also need this functionality. I guess for now I'm going to have to go with a janky workaround like a dummy `links` key in the `Cargo.toml`

**Comment by alex** [comment]
I'm another person with a similar use case (see: https://boringssl-review.googlesource.com/c/boringssl/+/63305).

I'd be more than happy to do some of the leg work to make this happen, what would the next step be there? Does this require an RFC, or just a design + PR?

**Comment by epage** [comment]
@davidben there is https://github.com/rust-lang/rust/issues/64797 though I think I heard they are wanting to start it with just `std`.  That is considered a blocker before considering adding rustc version `cfg`s.  I think we would likely want to go down the same route.  I'm assuming doing so would resolve that use case.  If so, I'd recommend creating an issue for that rather than hoping for this unrelated feature being extended to support it.

**Comment by Arnavion** [comment]
I use the same `links` hack in k8s-openapi since 2020 to let downstream libraries know which version feature of k8s-openapi was selected by the downstream-most binary crate, so that the libraries can do conditional compilation.

k8s-openapi supports multiple features like `v1_22`, `v1_23`, ... `v1_28`, corresponding to Kubernetes API server version. A binary crate makes the choice of which one of those it wants to enable. Any libraries between k8s-openapi and the binary crate that use k8s-openapi may need to compile differently based on which version feature was selected.

So k8s-openapi's build.rs emits `cargo:version=<value that encodes the enabled feature>` and downstream libraries can look at `DEP_ENV_K8S_OPENAPI_*_VERSION` env var to get that value and create their own `cfg`s, or whatever they want to do. https://arnavion.github.io/k8s-openapi/v0.20.x/k8s_openapi/#conditional-compilation

The alternative is that each such library exports its own version features that map one-to-one with the corresponding k8s-openapi features. This is busywork, especially when the chain of such crates is multiple crates long, or when a crate wants to match on a range of versions (eg all versions below 1.25). The latter would require reimplementing `<` using `cfg(any(feature = "v1_21", feature = "v1_22", ...))` which is also busywork and brittle. Also, if even a single crate doesn't do the re-exported features, or doesn't do it correctly, it ruins the whole thing.

`cfg(accessible())` or `cfg(version())` would not solve this problem in general. The specific example in my link (whether the `PodSpec::host_users` field exists or not) might be solved by `cfg(accessible())` , but there are other differences between the versions that cannot, like a field changing from type `Foo` to type `Option<Foo>`, or vice versa, or to a completely different type `Bar`, or to a type implementing `Default` in some versions but not others.

**Comment by epage** [comment]
> I use the same links hack in k8s-openapi since 2020 to let downstream libraries know which version feature of k8s-openapi was selected by the downstream-most binary crate, so that the libraries can do conditional compilation.

I wonder if https://internals.rust-lang.org/t/pre-rfc-mutually-excusive-global-features/19618 could help here.

> The specific example in my link (whether the PodSpec::host_users field exists or not) might be solved by cfg(accessible()) , but there are other differences between the versions that cannot, like a field changing from type Foo to type Option<Foo>, or vice versa, or to a completely different type Bar, or to a type implementing Default in some versions but not others.

Those sound like breaking changes which would imply controlling them with either system isn't a good fit.

**Comment by Arnavion** [comment]
>I wonder if https://internals.rust-lang.org/t/pre-rfc-mutually-excusive-global-features/19618 could help here.

As it is specified, no. Since the global values are limited to enums, it still has the problem of reimplementing `<` using `cfg(any(...))` I mentioned. But if it was extended to also support integers, that would indeed solve the problem.

>Those sound like breaking changes which would imply controlling them with either system isn't a good fit.

I don't understand what you mean.

**Comment by davidben** [comment]
> @davidben there is https://github.com/rust-lang/rust/issues/64797 though I think I heard they are wanting to start it with just std. That is considered a blocker before considering adding rustc version cfgs. I think we would likely want to go down the same route. I'm assuming doing so would resolve that use case. If so, I'd recommend creating an issue for that rather than hoping for this unrelated feature being extended to support it.

Oh, interesting. I guess I would call that unrelated mechanism for addressing a very related problem. But as long as it solves the problem, I'm happy. 😄 Tying it to symbol accessibility is interesting, but would work. Crates could define random constants that don't do anything useful, but are meant to be used with `#[cfg(accessible(...))]` queries. `accessible` is also a little verbose, but ah well.

I share @Arnavion's concern that `cfg` currently has no way to handle integers, which dramatically complicates things. In comparison, the C preprocessor, for all its warts, is perfectly capable of doing integer comparisons. Doing so is very, very common for version checks. The rust-openssl folks have a [whole mess](https://github.com/sfackler/rust-openssl/blob/master/openssl/build.rs) to work around this limitation.

But even booleans that don't require `build.rs` would be a dramatic improvement over the status quo. Still well behind C, but if extended to integers in the future, that would probably catch up.

**Comment by Arnavion** [comment]
>Crates could define random constants that don't do anything useful, but are meant to be used with `#[cfg(accessible(...))]` queries.

Oh, good point. This would also work k8s-openapi's case, eg enabling `v1_23` would define `pub const V1_23_ENABLED: ();`

So the lack of integers would be the only problem.

**Comment by epage** [comment]
> As it is specified, no. Since the global values are limited to enums, it still has the problem of reimplementing < using cfg(any(...)) I mentioned. But if it was extended to also support integers, that would indeed solve the problem.

Integer support is in the future possibilities.

**Comment by Arnavion** [comment]
I know. I understood "Future possibilities" to mean "not going to be in the RFC and initial implementation", so as it's currently specified it does not support integers.

**Comment by epage** [comment]
That is correct but the point more is that it has the potential for covering the case and it would be good to give feedback based on that possibility to see if it is going in the right direction or not.

**Comment by jschwe** [comment]
One other important use-case is when replacing a native library, with a rust implementation that still offers the C-FFI to be drop-in compatible, one still needs the C-headers that the library provides. But since the implementation is in rust, `links` suddenly doesn't make sense anymore, and one has to find another way to get the public header files to the consumer. 

One example being `icu-capi`, where the current recommendation for users is to use `cargo metadata` to figure out the on-disk location of header files, which brings in its own problems, like how to make the `cargo metadata` invocation in the build-script support both online and offline builds, without explicitly testing for network access.

**Comment by epage** [comment]
If you have Rust code depending on a Rust C FFI, doesn't that still need `links` for uniqueness and not just sharing data?

**Comment by jschwe** [comment]
Thats a good point! Yes, the library providing the C-API must be unique. Would your recommendation then be to just specify `links` anyway, even if the "native" library is implemented in rust, given that it still needs to be unique?

**Comment by epage** [comment]
Yes.  #5920 is the issue for uniqueness without `links`.

**Comment by Manishearth** [comment]
@epage AIUI `links` impacts a bunch of things: for a crate like `icu_capi` which does not link to native code I am very wary of using `links`, especially since we have ways of avoiding the symbol clash already. `icu_capi` does not actually need uniqueness within the same major version.

It also is typically interpreted differently by tooling that adapts Cargo crates to non-Cargo build systems.

**Comment by epage** [comment]
I was looking back over the entire thread, in particular
- https://github.com/rust-lang/cargo/issues/3544#issuecomment-302727633
- https://github.com/rust-lang/cargo/issues/3544#issuecomment-302850940


So putting those comments together, we would have:

Any build script can specify `cargo::metadata=<key>=<value>`
- `DEP_<links>_<key>=<value>` is only available if `package.links` is set
- `CARGO_DEP_<name_in_toml>_<key>=<value>` is always set

That seems relatively straight forward that I'm going to go ahead and mark this as `S-accepted` for being developed as an unstable feature.

Aside: places like https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts which refer to `DEP_<name>_<key>` seem like they should be changed to `DEP_<links>_<key>`.  See #16205.

Despite #14948, I wonder if should require this for #13663.  #13663 is about build scripts being able to generate final artifacts.  With `CARGO_DEP_<name_in_toml>_<key>=<value>`, we could limit staging to only user-selected packages which reduces the chance of dependencies doing something weird or of conflicts as the user is in charge of staging everything, rather than dependencies staging things.  I mentioned #14948 as this is requiring users to have more build scripts than they might not have needed it otherwise to forward artifacts along their dependency tree and we ideally would be reducing the number of build scripts needed.

**Comment by epage** [comment]
Because `DEP_<links>_<key>=<value>` remains unchanged throughout the dependency tree (because `<links>` is the same), `build-dependencies` can directly read it.  That won't be the case with `CARGO_DEP_<name_in_toml>_<key>=<value>` but people may assume it, overlooking renames.  We'll want to anchor the conversation in the docs that `build-dependencies` need to accept the env variable as input and not assume what it is to try to help.  There is little more we can do considering how sloppy env variables are.

**Comment by epage** [comment]
Another consideration in the design is that `DEP_LINKS_KEY=VALUE` [envifies](https://doc.rust-lang.org/nightly/nightly-rustc/src/cargo/core/compiler/mod.rs.html#1907-1912) `cargo::metadata=key=value`.  Skipping it makes the dependency name and key more predictable and greppable.

**Comment by ranger-ross** [comment]
#16436 is an attempt to implement the proposal in https://github.com/rust-lang/cargo/issues/3544#issuecomment-3474801325

**Comment by ranger-ross** [comment]
#16436 was merged and `-Zany-build-script-metadata` should land on nightly in about a week. 

`-Zany-build-script-metadata` allows any build script to pass env vars to dependent build scripts via `cargo::metadata=KEY=VALUE` even if it does not contain a `links` field in `Cargo.toml` as mentioned in [this comment](https://github.com/rust-lang/cargo/issues/3544#issuecomment-3474801325).

**Comment by LeonMatthes** [comment]
@ranger-ross Thank you for working on this issue.
Last year I worked on CXX-Qt, which also "abused" the links key to pass metadata between build scripts. I now work on Slint, where we're planning to use a similar system.

I want to clarify one aspect, which is the execution order of build scripts. IIRC, at the moment, if crate A is a normal dependency of B, this means the build scripts of A and B can be compiled and run in parallel, right?
Just the rustc invocation of A needs to run before B's.

Adding the links key to A currently changes this so that A's build script must complete before B's build script runs (to collect the metadata).
From what I can pick up from #16436, in future Cargo always completes A's build script before B's build script starts, whether a links key is set or not, correct?

If so, this is exactly what is needed for both CXX-Qt and Slint and I'm strongly in favor of stabilizing this 🙌

**Comment by ranger-ross** [comment]
> I want to clarify one aspect, which is the execution order of build scripts. IIRC, at the moment, if crate A is a normal dependency of B, this means the build scripts of A and B can be compiled and run in parallel, right? Just the rustc invocation of A needs to run before B's.
> 
> Adding the links key to A currently changes this so that A's build script must complete before B's build script runs (to collect the metadata). From what I can pick up from [#16436](https://github.com/rust-lang/cargo/pull/16436), in future Cargo always completes A's build script before B's build script starts, whether a links key is set or not, correct?

To my understanding, all of this is correct!

In #16436 this is changed, as we cannot know which build scripts will emit `cargo::metadata` statically. So build scripts will always wait for their dependencies's build scripts to finish to make sure the env vars are available and builds are reproducible.

**Comment by LeonMatthes** [comment]
Okay, great, then this feature as implemented in #16436 enables exactly the kind of cross-dependency build system integration that CXX-Qt and Slint require for a seamless Cargo workflow.
Thank you for working on this 🚀

**Comment by weihanglo** [comment]
One downside of [-Zany-build-script-metadata](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#any-build-script-metadata) is that build.rs starts depending on the crate's direct dependencies, which may make the Cargo build graph more linear.

https://github.com/rust-lang/cargo/blob/38623a0653b1ddc01f805fc85e14207c77b68178/src/cargo/core/compiler/unit_dependencies.rs?plain=1#L976-L980

I wonder whether this could instead be an opt-in feature on a per–build.rs basis.

That said, for build.rs that already has `package.links` set, it is already the case so nothing changes.

**Comment by ranger-ross** [comment]
> may make the Cargo build graph more linear.

For context that was discussed in this thread in the implementation PR
https://github.com/rust-lang/cargo/pull/16436#discussion_r2648502742
